### PR TITLE
Add information about node's original location in the parsed string

### PR DESCRIPTION
**Pull request by serso**
Prior to this commit it was not possible to trace back the parsed node location in the original string. This information could be used to highlight the syntax of a math expression displayed in a rich editor (e.g. a spreadsheet or a calculator).

This commit is based on the PR #2796 that has been abandoned for 2 years. I rebased the task branch and fixed the issues reported in the original PR.

In the nutshell, each parsed node stores an array of sources (`SourceMapping[]`) that are set during parsing (see `tokenSource` and its usages for details). The node's constructor and clone method are adjusted to take an optional `MetaOptions` object containing the source mappings. In the future `MetaOptions` could be extended to store more information.

Benchmarks showed no change in `evaluate`. `parse` became slower, from 3.68µs to 5.17µs with the changes from this commit.

Closes #2795

### CODE REVIEW

**Comment by serso** [comment]
@josdejong any chance you or someone else can look into this PR? Feedback is highly appreciated 🙏

**Comment by josdejong** [comment]
Hi Sergey, thanks a lot for picking this up! I think I don't have enough time today to review this (large) PR, I hope to look into it coming Friday.

File: docs/expressions/expression_trees.md [hunk 1]
```diff
@@ -39,10 +39,11 @@ tree generated by `math.parse('sqrt(2 + x)')`.
 
 All nodes have the following methods:
 
--   `clone() : Node`
+-   `clone(options: MetaOptions) : Node`
```
Comment (josdejong) [inline docs/expressions/expression_trees.md:42]: The `options` argument is optional, and it is named `meta` in the source code (consistent with the constructor arguments of the Node classes). Can you document this like `clone(meta?: MetaOptions) : Node`?

And similarly, the argument `meta` is optional in the constructors of each of the Node classes, can you udpate that too in this documentation page?

File: docs/expressions/expression_trees.md [hunk 2]
```diff
@@ -709,3 +728,25 @@ const node = math.parse('x')
...
+- `sources: SourceMapping` sets the sources for the newly created or cloned node
+
+### SourceMapping
+
+Each node has an array of `SourceMapping` objects which map back to the node's corresponding tokens in the original source string
+
+Properties:
+
+- `text: string` the token representing this node in the parsed string
+- `index; number` the index of the token in the parsed string
```
Comment (josdejong) [inline docs/expressions/expression_trees.md:751]: Small typo, I think the semicolon `;` sould be a colon `:`. 

All is all well documented BTW 👍

File: src/expression/node/Node.js [hunk 3]
```diff
@@ -7,6 +7,9 @@ import { createMap } from '../../utils/map.js'
 
 const name = 'Node'
 const dependencies = ['mathWithTransform']
+export const defaultMetaOptions = {
```
Comment (josdejong) [inline src/expression/node/Node.js:10]: It makes sense to define a `defaultMetaOptions`. However, would it make sense to create a new object and array every time to prevent people from being able to mess up by mutating this shared/reused object? Maybe we can use `Object.freeze` here to solve this. (Thinking aloud here)

File: src/expression/node/AccessorNode.js [hunk 4]
```diff
@@ -47,9 +48,10 @@ export const createAccessorNode = /* #__PURE__ */ factory(name, dependencies, ({
      * @param {Node} object                 The object from which to retrieve
      *                                      a property or subset.
      * @param {IndexNode} index             IndexNode containing ranges
+     * @param {MetaOptions} [meta]          The object with additional options for building this node.
      */
-    constructor (object, index) {
-      super()
+    constructor (object, index, meta = defaultMetaOptions) {
```
Comment (josdejong) [inline src/expression/node/AccessorNode.js:53]: Thinking aloud here: technically it is not needed to define a default value for `meta` in each of the Node classes since it will be set in the root Node.js class. What do you think? Would it be OK to leave `meta` as it is in these constructors and only set it in the constructor of `Node.js`? It would save a bit of overhead.

File: src/expression/node/Node.js [hunk 5]
```diff
@@ -7,6 +7,9 @@ import { createMap } from '../../utils/map.js'
...
+  sources: []
```
Comment (josdejong) [inline src/expression/node/Node.js:11]: I'm thinking about what would best communicate when `sources` are not available: an empty array like it is now? Or setting `sources: null` for example? In usage, it may be easier if you can know for sure that `sources` is always an array so you don't have to check whether it exists, so an empty array can make sense. But it can also be misleading and giving the impression that sources are available. Any thoughts?

File: src/expression/node/AccessorNode.js [hunk 6]
```diff
@@ -133,10 +135,11 @@ export const createAccessorNode = /* #__PURE__ */ factory(name, dependencies, ({
 
     /**
      * Create a clone of this node, a shallow copy
+     * @param {MetaOptions} [meta] An object with additional options for cloning this node
```
Comment (josdejong) [inline src/expression/node/AccessorNode.js:138]: Maybe a more clear explanation would be:

>  An object replacing the meta data of the cloned node

What do you think?

File: src/expression/parse.js [hunk 7]
```diff
@@ -214,6 +215,33 @@ export const createParse = /* #__PURE__ */ factory(name, dependencies, ({
...
   }
 
+  /**
+   * Returns a mapping of the current token in state to its place in the source expression
+   * @param {Object} state
+   * @return {SourceMapping} the source mapping
+   * @private
+   */
+  function tokenSource (state) {
+    if (config.traceSources) return { index: state.index - state.token.length, text: state.token }
```
Comment (josdejong) [inline src/expression/parse.js:225]: I think checking whether `config.traceSources` is true here is double work, it's already done at the place where this function is invoked by cheking whether `sources` is set:

```js
sources?.push(tokenSource(state))
```

Shall we remove the check from this function? Then we also cannot return an empty object which would be an invalid source.

File: src/expression/parse.js [hunk 8]
```diff
@@ -1304,11 +1377,14 @@ export const createParse = /* #__PURE__ */ factory(name, dependencies, ({
     if (state.tokenType === TOKENTYPE.SYMBOL && hasOwnProperty(state.extraNodes, state.token)) {
       const CustomNode = state.extraNodes[state.token]
 
+      const sources = config.traceSources ? [tokenSource(state)] : null
```
Comment (josdejong) [inline src/expression/parse.js:1380]: I see three ways to "initialize" sources/meta:

1. `config.traceSources ? [] : null`
2. `config.traceSources ? [tokenSource(state)] : null`
3. `tokenSourceMetaOptions(state)`, which basically does:
    `config.traceSources ? { sources: [tokenSource(state)] } : { sources: [] }`

I'm not sure, but can we unify this some way, maybe putting the `config.traceSources ? ... : ...` check in a single function that can be used for all three cases? We could replace the cases of `meta =` with `sources` is, and then at the end turn this in into an object via `{ sources }` instead of passing `meta` like is already the case in may places.

File: src/expression/parse.js [hunk 9]
```diff
@@ -1630,17 +1730,22 @@ export const createParse = /* #__PURE__ */ factory(name, dependencies, ({
...
+          sources?.push(tokenSource(state))
+        }
+
         getToken(state)
 
         if (state.token !== '}') {
           // parse key
           if (state.token === '"' || state.token === "'") {
-            key = parseStringToken(state, state.token)
+            key = parseStringToken(state, state.token).token
```
Comment (josdejong) [inline src/expression/parse.js:1748]: As far as I can see, the keys are not added to `sources`. Can we add them?

File: src/function/algebra/resolve.js [hunk 10]
```diff
@@ -65,24 +65,29 @@ export const createResolve = /* #__PURE__ */ factory(name, dependencies, ({
         nextWithin.add(node.name)
         return _resolve(value, scope, nextWithin)
       } else if (typeof value === 'number') {
-        return parse(String(value))
+        const parsed = parse(String(value)).clone({ sources: [] })
+        return parsed
       } else if (value !== undefined) {
-        return new ConstantNode(value)
+        const parsed = new ConstantNode(value).clone({ sources: [] })
```
Comment (josdejong) [inline src/function/algebra/resolve.js:71]: Why is `.clone` needed here?

File: src/function/algebra/resolve.js [hunk 11]
```diff
@@ -65,24 +65,29 @@ export const createResolve = /* #__PURE__ */ factory(name, dependencies, ({
...
+        return parsed
       } else {
         return node
       }
     } else if (isOperatorNode(node)) {
       const args = node.args.map(function (arg) {
         return _resolve(arg, scope, within)
       })
-      return new OperatorNode(node.op, node.fn, args, node.implicit)
+      const newNode = new OperatorNode(node.op, node.fn, args, node.implicit).clone({ sources: node.sources })
```
Comment (josdejong) [inline src/function/algebra/resolve.js:80]: Instead of using clone I think we can pass `meta` as argument to the constructor?

File: src/function/algebra/resolve.js [hunk 12]
```diff
@@ -65,24 +65,29 @@ export const createResolve = /* #__PURE__ */ factory(name, dependencies, ({
...
+      return newNode
     } else if (isParenthesisNode(node)) {
-      return new ParenthesisNode(_resolve(node.content, scope, within))
+      const parenNode = new ParenthesisNode(_resolve(node.content, scope, within)).clone({ sources: [] })
```
Comment (josdejong) [inline src/function/algebra/resolve.js:83]: Can we pass `meta` here too as argument to the constructor?

File: src/function/algebra/resolve.js [hunk 13]
```diff
@@ -65,24 +65,29 @@ export const createResolve = /* #__PURE__ */ factory(name, dependencies, ({
         nextWithin.add(node.name)
         return _resolve(value, scope, nextWithin)
       } else if (typeof value === 'number') {
-        return parse(String(value))
+        const parsed = parse(String(value)).clone({ sources: [] })
```
Comment (josdejong) [inline src/function/algebra/resolve.js:68]: It feels to me that using `parse` here is redundant and we can simply create `new ConstantNode(value, { sources: [] })`. But maybe I'm overlooking something? (Not really related to this PR but maybe we can improve this one line)

File: src/core/config.js [hunk 14]
```diff
@@ -33,5 +33,12 @@ export const DEFAULT_CONFIG = {
...
   // When false, it returns a matrix or array with a size depending on the type of index.
-  legacySubset: false
+  legacySubset: false,
+
+  // If set to `true` (the default value), `parse` records information about
+  // the original source location of each `Node` in the parsed string. See `SourceMapping` and `Node#sources`.
+  // If set to `false`, `Node#sources` is always be empty.
+  // The only time you want to set this to `false` is when you want to speed up parsing of
+  // a large amount of text.
+  traceSources: true
```
Comment (josdejong) [inline src/core/config.js:43]: I think `traceSources: false` may be a better default. Only in specific cases you need the sources, but peopler are mostly parsing an expression to just evaluate it straightaway. Since it slows down parsing and it is not needed in most use cases, it may be best to keep it turned off by default. What do you think?

File: test/unit-tests/expression/parse.test.js [hunk 15]
```diff
@@ -2605,4 +2615,228 @@ describe('parse', function () {
 
     assert.strictEqual(mathClone.evaluate('2'), 2)
   })
+
+  describe('sources', function () {
+    it('adds sources for constants', function () {
+      const parsed = math.parse('4')
+      assert.deepStrictEqual(parsed.sources, [{ index: 0, text: '4' }])
```
Comment (josdejong) [inline test/unit-tests/expression/parse.test.js:2622]: Can you test for example boolean and null? I see string is tested already lower down.

File: test/unit-tests/expression/parse.test.js [hunk 16]
```diff
@@ -2605,4 +2615,228 @@ describe('parse', function () {
...
+      // should have a source for brackets and item delimiters
+      const expected = [
+        { index: 0, text: '[' },
+        { index: 1, text: ']' }
+      ]
+
+      assert.deepStrictEqual(parsed.sources, expected)
+    })
+
+    it('adds sources for ranges', function () {
```
Comment (josdejong) [inline test/unit-tests/expression/parse.test.js:2692]: Can you add a test for a range like `A[2:end]` using the `end` symbol?

File: test/unit-tests/expression/parse.test.js [hunk 17]
```diff
@@ -2605,4 +2615,228 @@ describe('parse', function () {
...
+      const sources = [
+        { index: 3, text: '(' },
+        { index: 5, text: ',' },
+        { index: 8, text: ')' }
+      ]
+      assert.deepStrictEqual(parsed.sources, sources)
+    })
+
+    it('adds sources for string literals', function () {
+      const singleQuote = math.parse("'hello'")
```
Comment (josdejong) [inline test/unit-tests/expression/parse.test.js:2813]: I think in case of a string, the string value itself should also be an entry in the sources, right?

File: test/unit-tests/expression/parse.test.js [hunk 18]
```diff
@@ -2605,4 +2615,228 @@ describe('parse', function () {
...
+
+      const innerCond = outerCond.trueExpr.content
+      const innerSources = [
+        { index: 14, text: '?' },
+        { index: 18, text: ':' }
+      ]
+      assert.deepStrictEqual(innerCond.sources, innerSources)
+    })
+
+    it('adds sources for assignments', function () {
```
Comment (josdejong) [inline test/unit-tests/expression/parse.test.js:2749]: Can you also add tests for getting and setting (nested) properties on objects and arrays?

File: test/unit-tests/expression/parse.test.js [hunk 19]
```diff
@@ -2605,4 +2615,228 @@ describe('parse', function () {
...
+      assert.deepStrictEqual(parsedNull.sources, [{ index: 0, text: 'null' }])
+
+      const parsedInfinity = math.parse('Infinity')
+      assert.deepStrictEqual(parsedInfinity.sources, [{ index: 0, text: 'Infinity' }])
+
+      const parsedNaN = math.parse('NaN')
+      assert.deepStrictEqual(parsedNaN.sources, [{ index: 0, text: 'NaN' }])
+    })
+
+    it('adds sources for function calls', function () {
```
Comment (josdejong) [inline test/unit-tests/expression/parse.test.js:2800]: Can you add tests for defining a function (uses the `FunctionAssignmentNode`)?

File: test/unit-tests/expression/parse.test.js [hunk 20]
```diff
@@ -2605,4 +2615,228 @@ describe('parse', function () {
...
+      const parsed = math.parse('foo(1, 2)')
```
Comment (josdejong) [inline test/unit-tests/expression/parse.test.js:2801]: Can you add a test for a function call with more than two arguments?

File: test/unit-tests/expression/parse.test.js [hunk 21]
```diff
@@ -2605,4 +2615,228 @@ describe('parse', function () {
...
+
+      const doubleQuote = math.parse('"hello"')
+      const doubleSources = [
+        { index: 0, text: '"' },
+        { index: 6, text: '"' }
+      ]
+      assert.deepStrictEqual(doubleQuote.sources, doubleSources)
+    })
+
+    it('adds sources for objects', function () {
```
Comment (josdejong) [inline test/unit-tests/expression/parse.test.js:2828]: Can you add a test for an empty object?

File: src/core/function/config.js [hunk 22]
```diff
@@ -47,6 +47,8 @@ export function configFactory (config, emit) {
    *                            {string} randomSeed
    *                              Random seed for seeded pseudo random number generator.
    *                              Set to null to randomly seed.
+   *                            {boolean} traceSources
```
Comment (josdejong) [inline src/core/function/config.js:50]: Can you document this new configuration option `traceSources` on the page `docs/core/configuration.md`?

File: types/index.d.ts [hunk 23]
```diff
@@ -41,6 +41,17 @@ export interface FactoryFunctionMap {
   [key: string]: FactoryFunction<any> | FactoryFunctionMap
 }
 
+// Maps a parsed node back to its place in the original source string
+export interface SourceMapping {
```
Comment (josdejong) [inline types/index.d.ts:45]: Can you update the type definitions of the constructors of all Node classes and also their `.clone` method describing the new `meta` option?

**Review by josdejong — CHANGES_REQUESTED** [review]
Thanks again for picking this up Sergey! It's quite a job but it looks solid. it will be a very useful addition. 

I made quite some inline comments (nothing very big though). Can you have a look at those? Thanks!

**Comment by gwhitney** [comment]
Does this also supersede #2615 ?

**Comment by josdejong** [comment]
> Does this also supersede #2615 ?

Ah, yes. I'll close that PR in favor of this one. Thanks for pointing this out.

**Comment by josdejong** [comment]
@serso I just merged #3547 adding support for optional chaining. This causes a few merge conflicts with this PR. Please let me know if you need help resolving them.

**Comment by josdejong** [comment]
@serso can you have a look into the merge conflicts? Please let me know if you need help.
